### PR TITLE
Project wide occurrences

### PR DESCRIPTION
**Pull request by liam923**
This PR backports a large number of occurrence-related changes from upstream. This backport was done by:
1. Diffing `occurrences.ml` between Merlin-jst and upstream.
2. For each difference, attribute the difference to an upstream commit/PR (or a Merlin-jst specific consideration) and note the commit/PR.
3. Apply the upstream patches in chronological order. After each application, the testsuite was run. If tests failed, appropriate patches from upstream were found and applied.

After the backport, `occurrences.ml`'s only difference from upstream are Merlin-jst specific.

The backporting commit descriptions include the upstream PR or commit hash.

The first commit of this PR is not a backport but instead reverts fa1e8b182a0bc2f2d2db8b6e409e6378afab65a0. This is because is substantially differs from its upstream counterpart, 34e2688270306929111c25b87f60188da0335c5e. Thus, the decision was made to revert fa1e8b182a0bc2f2d2db8b6e409e6378afab65a0 and then cherry-pick 34e2688270306929111c25b87f60188da0335c5e.

There was a similar issue with 9e90e47f72f95c1741a7fdde16ea3f660503d5fa and 3c4d8588f97f72daf8394dfb554de64fe8d2ebe9, which are substantially different from the corresponding upstream commit 80f29ff907b79ceaeacd31788962fa720066ad85. This differences in `occurrences.ml` were fixed while merging. I may make a future PR to resolve other differences.

Note that this PR targets the branch `backport-shape-reduce`. That is becuase this PR builds on PR #78, which uses that branch.

Also note that the project-wide occurrences tests aren't run because ocaml-index hasn't yet been imported. I intend to import it and fix the tests in a follow-up PR.

### CODE REVIEW

File: src/analysis/ast_iterators.ml [hunk 1]
```diff
@@ -3,6 +3,90 @@ open Typedtree
 
 let {Logger. log} = Logger.for_section "iterators"
 
+(* Sometimes we do not want to iterate on nodes that do not correspond to actual
```
Comment (poechsel) [inline src/analysis/ast_iterators.ml:6]: This change seems to be unrelated to the backport

File: src/kernel/mpipeline.ml [hunk 2]
```diff
@@ -325,3 +336,30 @@ let timing_information t = [
...
+    | Miss -> `String "miss"
+    | Hit { reused; typed } ->
+      `Assoc
+        [ "reused" , `Int reused;
+          "typed", `Int typed
+        ]
+  in
+  let fmt_hit_miss h m =
+    `Assoc [ "hit", `Int h; "miss", `Int m ] in
+  let cmt_stat = Cmt_cache.get_cache_stats () in
```
Comment (poechsel) [inline src/kernel/mpipeline.ml:352]: Are the change made around caching needed? If yes we should think about making them work with cms files too

File: src/analysis/ast_iterators.ml [hunk 3]
```diff
@@ -45,7 +129,12 @@ let build_uid_to_locs_tbl ~(local_defs : Mtyper.typedtree) () =
   uid_to_locs_tbl
 
 let iter_on_usages ~f (local_defs : Mtyper.typedtree) =
-  let iter = Cmt_format.iter_on_occurrences ~f in
+  let occ_iter = Cmt_format.iter_on_occurrences ~f in
```
Comment (poechsel) [inline src/analysis/ast_iterators.ml:132]: Does this need any change to handle cms files?

Comment (liam923) [reply src/analysis/ast_iterators.ml:132]: No. `iter_on_occurrences` does not contain cmt-specific logic. In fact, `Cms_format` uses `Cmt_format.iter_on_occurrences`.

File: src/kernel/mpipeline.ml (hunk 2 above)
Comment (liam923) [reply src/kernel/mpipeline.ml:352]: I think yes - I added cms support here

**Comment by liam923** [comment]
Sorry I took a while to get to this but it's ready for re-review.

**Review by poechsel — APPROVED** [review]
